coralproject/talk
# Make regenerating story trees run as a job with a job status reporter API

## What does this PR do?

Moves the regenerate story trees functionality into a job processor queue.

Adds an api router at `<CORAL_URL>/api/jobStatus` that can be called with the `name`: `"regenerateStoryTrees` and return the progress of the story tree generation.

## What changes to the GraphQL/Database Schema does this PR introduce?

Adds optional `disableCommenting` and `disableCommentingMessage` fields to the `RegenerateStoryTreesInput` type.

## Does this PR introduce any new environment variables or feature flags? 

No new environment variables.

## If any indexes were added, were they added to `INDEXES.md`?

No new indexes.

## How do I test this PR?

- Start up Coral
- Create some stories or delete the `tree` field off of some stories
- Call the regenerate story trees graphql query to `<CORAL_URL>/api/graphql` with:
     ```
     mutation {
       regenerateStoryTrees(input: {
         clientMutationId: "1",
         disableCommenting: true,
         disableCommentingMessage: "doing story tree regeneration, closing comments while I think really hard"
       }) {
         accepted
         jobID
       }
     }
     ```
- Remember the `jobID` this returns, you'll need it in a moment
- Wait a minute or so, the job takes a bit to get going and start posting progress since it does a count to check how many stories it's going to update so it can report accurate progress
- Call `<CORAL_URL>/api/jobStatus` authenticated with an admin user token with the payload until you see that the percentage is `100`
    - NOTE: this is rate limited, so you can only call it twice a minute
    ```
    {
       "name": "regenerateStoryTrees"
       "jobID": "<YOUR_JOB_ID>"
    }
    ```
- Check your stories, see that they have correct `tree` fields generated for all stories under the tenant.
 
## How do we deploy this PR?

Requires the same deployment process at the 7.1.0 story tree migration since this is a fix to remedy the story tree generation issues in 7.1.0.
